### DISCUSSION TITLE
feat: add service manager

### DISCUSSION THREAD

**Discussion by vgonkivs** [celestiaorg/celestia-node#2986] [Ideas]
Some time ago we discussed with @Wondertan that at some point we'll need some kind of service that can manage our services. 
The first use case, that it could handle is simplifying our [ServiceBreaker](https://github.com/celestiaorg/celestia-node/blob/31f0e25a89e229620d14b15f9a22f4b8ec2a20f8/nodebuilder/fraud/lifecycle.go#L23). TLDR: ServiceBreaker wraps some of our services and stops them once BEFP appeared. But it has one disadvantage - each service has to create its pubsub subscription over the BEFP topic. But we can have a service that creates a single subscription, waits for the BEFP, and then stops all respective services. This is only one use case that I can imagine rn, but I'm sure we can find more in the future. 

```go
type Service interface {
    Attach(topic string, stop func(context.Context) error) error
    Detach(topic string) error
}
```

**Comment by renaynay** [comment]
i'm fine with this approach, let's make it internal. Not high prio but would help clean up nodebuilder for sure. Thoughts on making this a good first issue ?

**Reply by vgonkivs** [reply]
I don't mind making it a good first issue. Should I prepare a detailed step-by-step explanation of what should be done? cc @Wondertan

**Reply by vgonkivs** [reply]
But I'd also propose to make an EPIC for this because the task can be decoupled

**Reply by Wondertan** [reply]
I don't think this is something a newcomer can handle. I would visit the manager once we have time to do refactorings on the nodebuilder.

**Comment by Wondertan** [comment]
Is this targeted particularly for managing PubSub subscriptions?

**Reply by vgonkivs** [reply]
At this point, yes. But I'm opened to any suggestions. We can make it configurable to manage whatever
